engine/runtime/src/main/java/org/enso/interpreter/runtime/error/WithWarnings.java

```diff
@@ -160,15 +163,31 @@ public int hashCode(Object o) {
   @CompilerDirectives.TruffleBoundary
   private EconomicSet<Warning> createSetFromArray(Warning[] entries) {
     EconomicSet<Warning> set = EconomicSet.create(new WarningEquivalence());
-    set.addAll(Arrays.stream(entries).iterator());
+    for (int i=0; i<entries.length; i++) {
+      if (set.size() >= MAX_WARNINGS_LIMIT) {
+        return set;
+      }
+      set.add(entries[i]);
+    }
+
     return set;
   }
 
   @CompilerDirectives.TruffleBoundary
   private EconomicSet<Warning> cloneSetAndAppend(EconomicSet<Warning> initial, Warning[] entries) {
     EconomicSet<Warning> set = EconomicSet.create(new WarningEquivalence());
-    set.addAll(initial.iterator());
-    set.addAll(Arrays.stream(entries).iterator());
+    for (Warning warning: initial) {
+      if (set.size() >= MAX_WARNINGS_LIMIT) {
```

Comment:
First of all `initial` should never be higher than the `MAX_WARNING_LIMIT`, right?

Disappointing to still see this whole manipulation behind a `@TruffleBoundary`...

...to make some performance improvement: can we check the `initial` is already at the `MAX_WARNINGS_LIMIT` and in such case just return `initial`? Such a check should be doable outside of `@TruffleBoundary`, so for saturated warnings we could be faster.

Reply:
> Disappointing to still see this whole manipulation behind a @TruffleBoundary...

I know that a unicorn died but this wasn't the task of this PR. Agreed with @Akirathan that I'll be waiting for the improved `EnsoHashMap` and then it will be easier to do something about it.

Reply:
> Disappointing to still see this whole manipulation behind a @TruffleBoundary...

Tracked in #5233. It will be, indeed, done later once we have PE-friendly `EnsoHashMap`. I have it in my backlog.